File: 0150-evaluate-reverse-polish-notation/0150-evaluate-reverse-polish-notation.js
/**
 * @param {string[]} tokens
 * @return {number}
 */
var evalRPN = function(tokens) {
  const stack = [];

  for (let token of tokens) {
    const parsedNum = Number(token);

    if (!Number.isNaN(parsedNum)) {
      stack.push(parsedNum);
    }
    else {
      const operand2 = stack.pop();
      const operand1 = stack.pop();
      switch (token) {
        case '+':
          stack.push(operand1 + operand2);
          break;
        case '-':
          stack.push(operand1 - operand2);
          break;
        case '*':
          stack.push(operand1 * operand2);
          break;
        case '/':
          stack.push(Math.trunc(operand1 / operand2));
          break;
      }
    }
  }

  if (stack.length !== 1) throw new Error("invalid RPN");

  return stack[0];
};


console.log(evalRPN(["2","1","+","3","*"]), 'should be 9');

File: 0150-evaluate-reverse-polish-notation/0150-evaluate-reverse-polish-notation.py
class Solution:
    def evalRPN(self, tokens: list[str]) -> int:
        stack = []

        for token in tokens:
            match token:
                case '+':
                    stack.append(stack.pop() + stack.pop())
                case '-':
                    op2, op1 = stack.pop(), stack.pop()
                    stack.append(op1 - op2)
                case '*':
                    stack.append(stack.pop() * stack.pop())
                case '/':
                    op2, op1 = stack.pop(), stack.pop()
                    stack.append(int(op1 / op2))
                case _:
                    stack.append(int(token))

        return stack[0]
